src/index.js

```diff
@@ -89,7 +89,19 @@ export default class VueRouter {
 
     this.apps.push(app)
 
-    // main app already initialized.
+    // set up app destroyed handler
+    // https://github.com/vuejs/vue-router/issues/2639
+    app.$once('hook:destroyed', () => {
+      // clean out app from this.apps array once destroyed
+      const index = this.apps.indexOf(app)
+      if (index > -1) this.apps.splice(index, 1)
+      // ensure we still have a main app or null if no apps
+      // we do not release the router so it can be reused
+      if (this.app === app) this.app = this.apps[0] || null
+    })
+
+    // main app previously initialized
+    // return as we don't need to set up new history listener
     if (this.app) {
       return
     }
```

Comment:
With `this.app` being set to null (if all apps are cleared out), this early return doesn't happen, and the code below it will be executed again.

Will that be an issue? IF so, maybe a flag needs to be set so that the following code is not executed a second time.

Reply:
but this is called when a `router` is injected in a created component, it's only executed once per root instance

Reply:
Doesn't `init` get called for each app instance, as it is how the apps get added to `this.apps` array:

```js
  init (app: any /* Vue component instance */) {
    process.env.NODE_ENV !== 'production' && assert(
      install.installed,
      `not installed. Make sure to call \`Vue.use(VueRouter)\` ` +
      `before creating root instance.`
    )

    this.apps.push(app)
```

Reply:
yes, but once per root instance

Reply:
Ah, correct... so it should be good.